### PR TITLE
Add conditional exit after control point.

### PR DESCRIPTION
**Pull request by ptersilie**
Companion PR: https://github.com/ykjit/yk/pull/500

Since the stopgap interpreter may interpret through the exit from the
main interpreter loop, we can't just continue running from the control
point after it returns. The stopgap interpreter now tells us via its
return value whether we need to continue as normal, or exit the main
interpreter loop. This is currently done by returning a boolean from the
control point, where `true` means the stopgap interpreter has run out of
the main interpreter loop. In this case we simply need to return.

Note, that we actually also need to return the value of the return that
was interpreted by the stopgap interpreter. We currently don't do this
since it requires more changes to the control point (passing in a
pointer in which to store the return value).

### CODE REVIEW

File: llvm/lib/Transforms/Yk/ControlPoint.cpp [hunk 1]
```diff
@@ -173,7 +173,7 @@ class YkControlPoint : public ModulePass {
 
     // Create the new control point.
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/ControlPoint.cpp:174]: Can we add a comment giving an idea of the type signature of the control point we're creating?

Comment (ptersilie) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:174]: Sure. I'm not too sold on the current design (though it does keep us going). The control point currently returns a bool to tell the interpreter whether to continue, or return, which is a bit crude. We also need to return the actual return value via the control point which we are currently not doing. Two options I can think of: a) return a struct, e.g.  `{retval, continue}`, or pass one of them into the control point as a pointer, which the stopgap interpreter then writes to.

File: llvm/lib/Transforms/Yk/ControlPoint.cpp [hunk 2]
```diff
@@ -222,6 +223,28 @@ class YkControlPoint : public ModulePass {
...
     OldCtrlPointCall->eraseFromParent();
 
+    // Get the result of the control point call. If it returns true, that means
+    // the stopgap interpreter has interpreted a return so we need to return as
+    // well.
+    
+    // Create the new exit block.
+    BasicBlock *ExitBB = BasicBlock::Create(Context, "", Caller);
+    Builder.SetInsertPoint(ExitBB);
+    Builder.CreateRet(ConstantInt::get(Type::getInt32Ty(Context), 0)); // XXX need to return value of interpreted return
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/ControlPoint.cpp:233]: s/XXX/FIXME/

File: llvm/lib/Transforms/Yk/ControlPoint.cpp (hunk 1 above)
Comment (ltratt) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp:174]: I'm fine with this as an interim design, while we work out if it's general enough or not. I just would like us to document what the control point will "look like" (so to speak) so our future selves don't have to reverse engineer all the LLVM function calls :)

**Comment by ptersilie** [comment]
Unfortunately, I just noticed that I forgot to force push before raising this PR. So I can't push my changes unless I force-push now. Asking for permission and forgiveness.

**Comment by ltratt** [comment]
Can you rebase against this branch as it is?

**Comment by ptersilie** [comment]
Not sure. I've made a change, squashed it, and then (without pushing) raised this PR.

**Comment by ltratt** [comment]
You can use `git reflog` to undo the squash and then there'll be no need for a force push. Search for e.g. "git unsquash" online.

**Comment by ptersilie** [comment]
Ok, I used `reflog` to get unsquash and then cherry picked my new changes on top. Not sure it was worth it for this minor change, but at least I've learned something new.

File: llvm/lib/Transforms/Yk/ControlPoint.cpp (hunk 1 above)
Comment (ptersilie) [reply llvm/lib/Transforms/Yk/ControlPoint.cpp]: Fixed in [eda3b57](https://github.com/ykjit/ykllvm/pull/22/commits/eda3b57f4570e970def8c4138f4a9e73988ceb20)

**Comment by ptersilie** [comment]
Can I squash them?

**Comment by vext01** [comment]
LGTM

**Comment by ptersilie** [comment]
Yeah, might as well.

**Comment by ltratt** [comment]
You need to do a rebase as well as a squash. Please force push an update.

**Comment by vext01** [comment]
What's the status of this @ptersilie?

**Comment by ptersilie** [comment]
I'm currently trying to get the LLVM tests to pass.

**Comment by ptersilie** [comment]
This is ready for bors now.

**Comment by ltratt** [comment]
I see a force push -- what changed?

**Comment by vext01** [comment]
According to the "compare changes" button, nothing: https://github.com/ykjit/ykllvm/compare/4370dc73f68534f1e2b4a1969eb9fd1b0fe21fd8..a10ba007ccf275319385de2248e433962f65170f

**Comment by ptersilie** [comment]
> I see a force push -- what changed?

I had to sign the commit. I assumed I still had permission since nothing else changed since you told me to squash and rebase.

**Comment by ltratt** [comment]
Not ideal, so pretty please let's not do this for the next PR.

bors r+

**Pull request merged**